src/libcrun/handlers/wamr.c

```diff
@@ -0,0 +1,242 @@
+    error (EXIT_FAILURE, 0, "could not find wasm_runtime_get_wasi_exit_code symbol in `libiwasm.so`");
+  if (wasm_application_execute_main == NULL)
+    error (EXIT_FAILURE, 0, "could not find wasm_application_execute_main symbol in `libiwasm.so`");
+  if (wasm_runtime_set_wasi_args == NULL)
+    error (EXIT_FAILURE, 0, "could not find wasm_runtime_set_wasi_args symbol in `libiwasm.so`");
+
+  int ret;
+  const char *exception;
+  cleanup_free char *buffer = NULL;
+  char error_buf[128];
+  size_t buffer_size;
+  uint32_t stack_size = 8096, heap_size = 0;
+  libcrun_error_t tmp_err = NULL;
+  const char *wasi_proc_exit_exception = "wasi proc exit";
+
+  const char *dirs[2] = { "/", "." };
+  char **container_env = container->container_def->process->env;
+  size_t env_count = container->container_def->process->env_len;
+
+  int arg_count = 0;
+  char *const *arg;
+  for (arg = argv; *arg != NULL; ++arg)
+    arg_count++;
+
+  // initialize the wasm runtime by default configurations
+  if (! wasm_runtime_init ())
+    error (EXIT_FAILURE, 0, "Failed to initialize the wasm runtime");
+
+  // read WASM file into a memory buffer
+  ret = read_all_file (pathname, &buffer, &buffer_size, &tmp_err);
+  if (UNLIKELY (ret < 0))
+    {
+      crun_error_release (&tmp_err);
+      error (EXIT_FAILURE, 0, "Failed to read file");
+    }
+
+  if (UNLIKELY (buffer_size > UINT32_MAX))
+    error (EXIT_FAILURE, 0, "File size is too large");
+
+  // parse the WASM file from buffer and create a WASM module
+  module = wasm_runtime_load (buffer, buffer_size, error_buf, sizeof (error_buf));
+  if (! module)
+    error (EXIT_FAILURE, 0, "Failed to load WASM file");
+
+  // instantiate the WASI environment
+  wasm_runtime_set_wasi_args (module, dirs, 1, NULL, 0, (const char **) container_env, env_count, (char **) argv, arg_count);
+
+  // create an instance of the WASM module (WASM linear memory is ready)
+  module_inst = wasm_runtime_instantiate (module, stack_size, heap_size, error_buf, sizeof (error_buf));
+  if (! module_inst)
+    error (EXIT_FAILURE, 0, "Failed to instantiate the WASM module");
+
+  // look up a WASM function by its name (The function signature can NULL here)
+  func = wasm_runtime_lookup_function (module_inst, "_start");
+  if (! func)
+    error (EXIT_FAILURE, 0, "Failed to look up the WASM function");
+
+  // create an execution environment to execute the WASM functions
+  exec_env = wasm_runtime_create_exec_env (module_inst, stack_size);
+  if (! exec_env)
+    error (EXIT_FAILURE, 0, "Failed to create the execution environment");
+
+  // call the WASM function
+  ret = wasm_runtime_call_wasm (exec_env, func, 0, NULL);
+  if (ret)
+    wasm_runtime_set_exception (module_inst, wasi_proc_exit_exception);
+  exception = wasm_runtime_get_exception (module_inst);
+  if (! strstr (exception, wasi_proc_exit_exception))
+    error (EXIT_FAILURE, 0, "Failed to call the WASM function");
+  wasm_runtime_clear_exception (module_inst);
```

Comment:
Wouldn't it be more straight-forward to just use wasm_runtime_get_exception() instead of first using wasm_runtime_set_exception() and then wasm_runtime_get_exception() ? Also why strstr() instead of strcmp()? 
I haven't had time to take a closer look so this more of a drive-by comment.